src/twisted/internet/asyncfs.py

```diff
@@ -0,0 +1,709 @@
+         - C{disk_namemax} maximum path length (int)
+         - C{disk_label} disk/volume label (str)
+         - C{disk_fstype} filesystem type (str)
+         - C{disk_birthtime} disk/volume creation time (int)
+
+        @rtype: L{dict}
+        """
+
+
+class IFile(Interface):
+    """
+    This represents an open file on the filesystem.  An object adhering to this
+    interface should be returned from L{openFile}().
+    """
+
+    def close() -> "Deferred":
+        """
+        Close the file.
+
+        This method returns nothing if the close succeeds immediately, or a
+        Deferred that is called back when the close succeeds.
+        """
+
+    def readChunk(offset: int, length: int) -> "Deferred":
+        """
+        Read from the file.
+
+        If EOF is reached before any data is read, raise EOFError.
+
+        This method returns the data as L{bytes}, or a Deferred that is
+        called back with same.
+
+        @param offset: an integer that is the index to start from in the file.
+        @param length: the maximum length of data to return.  The actual amount
+        returned may less than this.  For normal disk files, however,
+        this should read the requested number (up to the end of the file).
+        """
+
+    def writeChunk(offset: int, data: bytes) -> "Deferred":
+        """
+        Write to the file.
+
+        This method returns when the write completes, or a Deferred that is
+        called when it completes.
+
+        @param offset: an integer that is the index to start from in the file.
+        @param data: a L{bytes} that is the data to write.
+
+        @return: number of bytes written
+        @rtype: L{int}
+        """
+
+    def getAttrs() -> "Deferred":
+        """
+        Return the attributes for the file.
+
+        This method returns a dictionary in the same format as the attrs
+        argument to L{openFile} or a L{Deferred} that is called back with same.
+        """
+
+    def setAttrs(attrs: dict) -> "Deferred":
+        """
+        Set the attributes for the file.
+
+        This method returns when the attributes are set or a Deferred that is
+        called back when they are.
+
+        @param attrs: a dictionary in the same format as the attrs argument to
+        L{openFile}.
+        """
+
+    def fileno() -> int:
+        """
+        @return: the underlying file descriptor
+        @rtype: L{int)
+        """
+
+    def send(
+        consumer: "IConsumer", start: int = 0, chunkSize: int = DEFAULT_CHUNK_SIZE
+    ) -> "Deferred":
+        """
+        Produce the contents of the file to the given consumer.
+
+        @type consumer: C{IConsumer}
+        @param start: offset to begin reading from
+        @param chunkSize: size, in bytes, of each read
+        @return: A Deferred which fires when the file has been
+        consumed completely.
+
+        adapted from L{twisted.protocols.ftp.IReadFile}
+        """
```

Comment:
Depending on various implementation details, I think it will probably make the most sense to implement `readChunk` in terms of send.  If `send` also accepts a length then all `readChunk` does is adapt the output/return type to a single string (from a bunch of writes on a consumer).  The underlying implementation should otherwise be identical (and therefore should just be the same).

This also means `readChunk` can come for free for all implementations that provide `send`.

Reply:
I would prefer the other way round as the reading primitive on `io_uring` (and `os` for threads) is like `readChunk`. For the `io_uring` implementation (and probably others) `send` I would implement in `readChunk`.  This could be done for threads also but I assumed (possibly wrongly) that a single long-running task was important for efficiency.

Reply:
Twisted's threading primitives are pool-based so there's little or no difference between a long-lived task and N short-lived tasks from an efficiency perspective.  Either way you have to pass messages between the reactor thread and a worker thread.  There _is_ a difference in contention for threads in the pool though.  If you have long-lived tasks hogging all of the threads in the pool then no short-lived tasks can run until some long-lived tasks finish.

So it's preferable to have short-lived tasks in default/global thread pool.

If there is an efficiency argument to be made (and maybe there *shouldn't* be - at least not without benchmarks), it is probably that the interface should not involved the creation of an additional Deferred for every piece of data that moves through the system because creating, using, and destroying a Deferred is significantly more expensive than shoveling a `bytes` object from one place to another.  With `send` I can make anywhere from 1 to N Deferreds by choosing how much data to work on but *without* committing myself to holding all of that data in memory at once.  With `writeChunk` I have to balance number of Deferreds created against amount of data held in memory at a time.

If there is some argument based on the underlying system primitives (ie `io_uring` or Windows IOCP) then perhaps `send` and `writeChunk` should be on separate interfaces.  An implementer can choose whichever best fits the underlying abstraction and we can provide adapters from each to the other for use when one interface or the other is better suited to the application code (although speaking of application code, I wonder if txtubes is the *actually* correct application-facing interface).